File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/main/settings/SettingsScreen.java
package opencrowd.hgc.hgcwallet.ui.main.settings;


import android.content.Context;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.CompoundButton;
import android.widget.RadioButton;
import android.widget.TextView;

import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.common.UserSettings;
import opencrowd.hgc.hgcwallet.ui.auth.AuthActivity;
import opencrowd.hgc.hgcwallet.local_auth.AuthListener;
import opencrowd.hgc.hgcwallet.local_auth.AuthManager;
import opencrowd.hgc.hgcwallet.local_auth.AuthType;
import opencrowd.hgc.hgcwallet.ui.auth.fingerprint.FingerprintActivityHelper;
import opencrowd.hgc.hgcwallet.ui.customviews.TitleBarWrapper;

public class SettingsScreen extends Screen<SettingsScreenView> implements AuthListener {
    @Override
    protected SettingsScreenView createView(@NonNull Context context) {
        return new SettingsScreenView(context);
    }

    @Override
    public void onAuthSetupSuccess() {

    }

    @Override
    public void onAuthSuccess(int requestCode) {

    }

    @Override
    public void onAuthFailed(int requestCode, boolean isCancelled) {

    }

    @Override
    public void onAuthSetupFailed(boolean isCancelled) {
        SettingsScreenView view = getView();
        if (view != null)
            view.reloadUI();
    }
}

class SettingsScreenView extends BaseScreenView<SettingsScreen> {

    private TextView mEdittextName;
    private RadioButton mPinSetup;
    private RadioButton mFingerprintSetup;

    public SettingsScreenView(@NonNull Context context) {
        super(context);
        inflate(context, R.layout.view_settings_layout,this);
        TitleBarWrapper titleBar = new TitleBarWrapper(findViewById(R.id.titleBar));
        titleBar.setTitle(R.string.text_profile);
        mEdittextName = findViewById(R.id.edittext_name);
        mFingerprintSetup = findViewById(R.id.btn_radio_fingerprint);
        mPinSetup = findViewById(R.id.btn_radio_pinsetup);
        reloadUI();

        mEdittextName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                UserSettings.instance.setValue(UserSettings.KEY_USER_NAME,mEdittextName.getText().toString());
            }
        });

        mFingerprintSetup.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (b && AuthManager.INSTANCE.getAuthType() != AuthType.FINGER) {
                    FingerprintActivityHelper helper = new FingerprintActivityHelper();
                    helper.setup(getContext(), true, null);
                    if (helper.errorMsg == null) {
                        AuthActivity authActivity = (AuthActivity) getScreen().getActivity();
                        authActivity.setupAuth(AuthType.FINGER);

                    } else {
                        reloadUI();
                        Singleton.INSTANCE.showToast(getScreen().getActivity(),helper.errorMsg);
                    }
                }
            }
        });

        mPinSetup.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (b && AuthManager.INSTANCE.getAuthType() != AuthType.PIN) {
                    AuthActivity authActivity = (AuthActivity) getScreen().getActivity();
                    authActivity.setupAuth(AuthType.PIN);
                }
            }
        });
    }

    public void reloadUI() {
        mFingerprintSetup.setChecked(AuthManager.INSTANCE.getAuthType() == AuthType.FINGER);
        mPinSetup.setChecked(AuthManager.INSTANCE.getAuthType() == AuthType.PIN);
        String name = UserSettings.instance.getValue(UserSettings.KEY_USER_NAME);
        if(name != null && !name.isEmpty()) {
            mEdittextName.setText(name);
        }
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/main/settings/SettingsScreen.kt
package hedera.hgc.hgcwallet.ui.main.settings

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen
import hedera.hgc.hgcwallet.R
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.common.UserSettings
import hedera.hgc.hgcwallet.database.wallet.Wallet
import hedera.hgc.hgcwallet.local_auth.AuthListener
import hedera.hgc.hgcwallet.local_auth.AuthManager
import hedera.hgc.hgcwallet.local_auth.AuthType
import hedera.hgc.hgcwallet.ui.LauncherActivity
import hedera.hgc.hgcwallet.ui.auth.AuthActivity
import hedera.hgc.hgcwallet.ui.auth.fingerprint.FingerprintActivityHelper
import hedera.hgc.hgcwallet.ui.customviews.TitleBarWrapper
import hedera.hgc.hgcwallet.ui.onboard.EmptyScreen
import java.lang.Exception

class SettingsScreen : Screen<SettingsScreenView>(), AuthListener {


    data class Params(val name: String?, val defaultFee: String, var authType: AuthType)

    private val params = Params(UserSettings.getValue(UserSettings.KEY_USER_NAME), Singleton.format(Singleton.toCoins(Singleton.getDefaultFee())), AuthManager.authType)

    override fun createView(context: Context): SettingsScreenView {
        return SettingsScreenView(context, params)
    }

    override fun onAuthSetupSuccess() {
        params.authType = AuthManager.authType
        view?.reloadUI()
    }

    override fun onAuthSuccess(requestCode: Int) {

    }

    override fun onAuthSetupFailed(isCancelled: Boolean) {}

    override fun onAuthFailed(requestCode: Int, isCancelled: Boolean) {
        view?.reloadUI()
    }

    internal fun onBiometricClick() {
        if (AuthManager.authType != AuthType.FINGER) {
            val helper = FingerprintActivityHelper()
            helper.setup(activity, true, null)
            if (helper.errorMsg == null)
                (activity as? AuthActivity)?.setupAuth(AuthType.FINGER)
            else
                Singleton.showToast(activity, helper.errorMsg)

        }
    }

    internal fun onPinClick() {
        (activity as? AuthActivity)?.setupAuth(AuthType.PIN)
    }

    override fun onHide(context: Context?) {
        super.onHide(context)
        view?.getDefaultFee()?.let {
            onFeeChange(it)
        }
    }

    internal fun onFeeChange(text: CharSequence?) {
        val result = text?.toString()?.toDoubleOrNull()
        if (result == null)
            activity?.run { Singleton.showDefaultAlert(this, getString(R.string.invalid_fee_title), getString(R.string.invalid_fee_message)) }
        else {

            UserSettings.setValue(UserSettings.KEY_DEFAULT_FEE, Singleton.toNanoCoins(result))
        }
    }
}


class SettingsScreenView(context: Context, val params: SettingsScreen.Params) : BaseScreenView<SettingsScreen>(context) {

    private val edittextName: TextView?
    private val edittextDefaultFee: EditText?
    private val btnPin: Button?
    private val btnFingerprint: Button?

    init {
        View.inflate(context, R.layout.view_settings_layout, this)

        TitleBarWrapper(findViewById(R.id.titleBar)).apply {
            setTitle(R.string.text_profile)
        }

        edittextName = findViewById<TextView>(R.id.edittext_name)?.apply {
            hint = "Name account locally"
            addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                }

                override fun afterTextChanged(s: Editable) {
                    text?.toString()?.let { UserSettings.setValue(UserSettings.KEY_USER_NAME, it) }
                }

            })
        }

        edittextDefaultFee = findViewById<EditText>(R.id.edittext_default_fee)


        btnFingerprint = findViewById<Button>(R.id.btn_fingerprint)?.apply {
            setOnClickListener { screen?.onBiometricClick() }
        }

        btnPin = findViewById<Button>(R.id.btn_pin)?.apply {
            setOnClickListener { screen?.onPinClick() }
        }


        reloadUI()


    }


    fun reloadUI() {

        if (!params.name.isNullOrEmpty())
            edittextName?.text = params.name

        edittextDefaultFee?.setText(params.defaultFee)

        btnPin?.visibility = View.VISIBLE
        btnFingerprint?.visibility = View.VISIBLE

        when (params.authType) {
            AuthType.PIN -> {
                btnPin?.setText(R.string.change_pin)
                btnFingerprint?.setText(R.string.enable_biometric)
            }
            AuthType.FINGER -> {
                btnPin?.setText(R.string.enable_pin)
                btnFingerprint?.visibility = View.GONE
            }
            AuthType.UNKNOWN -> {
                btnPin?.setText(R.string.enable_pin)
                btnFingerprint?.setText(R.string.enable_biometric)
            }
        }
    }

    fun getDefaultFee() = edittextDefaultFee?.text.toString()

}